### PR TITLE
Feature/open source

### PR DESCRIPTION
**Pull request by mattbaumann1**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,5 +2,82 @@
...
-This repository contains a re-factored version of the [NITRO-Competitions](https://github.com/NUBIC/nitro-competitions) original code base which is currently running at Northwestern University Clinical and Translational Institute (NUCATS). It is work-progress, and intended primarily to enable a cloud-based implementtion in support of the [CTSA National Center for Data to Health (CD2H)](https://ctsa.ncats.nih.gov/cd2h/) operations. 
+This repository contains a re-factored version of the [NITRO-Competitions](https://github.com/NUBIC/nitro-competitions) original code base which is currently running at Northwestern University Clinical and Translational Institute (NUCATS).
+With the support of the [CTSA National Center for Data to Health (CD2H)](https://ctsa.ncats.nih.gov/cd2h/), the new application supports a cloud-based implementation and SAML authentication.
 
+# Prerequisites
+
+Installing Competitions requires certain configurations.
+  * A shibboleth IDP for these configuration
+  * Storage should be configured locally or with AWS.
+    - It is possible to user other services, but more changes will be necessary.
```
Comment (jseraf) [inline README.md:13]: These lines seem more related to the `### config/storage.yml` section in Installation.

File: README.md [hunk 2]
```diff
@@ -2,5 +2,82 @@
 
 Competitions is an open source tool to run NIH-style peer review of competitions, pilot projects, and research proposals in a cloud-based consortium-wide single sign-on platform.
 
-This repository contains a re-factored version of the [NITRO-Competitions](https://github.com/NUBIC/nitro-competitions) original code base which is currently running at Northwestern University Clinical and Translational Institute (NUCATS). It is work-progress, and intended primarily to enable a cloud-based implementtion in support of the [CTSA National Center for Data to Health (CD2H)](https://ctsa.ncats.nih.gov/cd2h/) operations. 
+This repository contains a re-factored version of the [NITRO-Competitions](https://github.com/NUBIC/nitro-competitions) original code base which is currently running at Northwestern University Clinical and Translational Institute (NUCATS).
+With the support of the [CTSA National Center for Data to Health (CD2H)](https://ctsa.ncats.nih.gov/cd2h/), the new application supports a cloud-based implementation and SAML authentication.
 
+# Prerequisites
+
+Installing Competitions requires certain configurations.
```
Comment (jseraf) [inline README.md:10]: The following are required to install competitions:

File: README.md [hunk 3]
```diff
@@ -2,5 +2,82 @@
...
+  * A shibboleth IDP for these configuration
+  * Storage should be configured locally or with AWS.
+    - It is possible to user other services, but more changes will be necessary.
+  * Mailcatcher(https://mailcatcher.me) is used in the development environment for emails.
+  * Postgres database
+  * RVM installed on server
```
Comment (jseraf) [inline README.md:16]: Maybe put Ruby version here? 

Is RVM a requirement? Could use [another ruby manager](https://www.ruby-lang.org/en/documentation/installation/#managers)

File: README.md [hunk 4]
```diff
@@ -2,5 +2,82 @@
...
+  service: S3
+  access_key_id: <AWS_ACCESS_KEY_ID>
+  secret_access_key: <AWS_SECRET_ACCESS_KEY>
+  region: <REGION_FOR_AWS_BUCKET>
+  bucket: <NAME_OF_YOUR_BUCKET>
+```
+
+### config/environments
+
+Included are a test and development environment configured out of the box. There are also example environment files for staging and production included. These will need the following configured before they will work.
```
Comment (jseraf) [inline README.md:74]: `Test and development environments are configured out of the box.`

File: README.md [hunk 5]
```diff
@@ -2,5 +2,82 @@
...
 Competitions is an open source tool to run NIH-style peer review of competitions, pilot projects, and research proposals in a cloud-based consortium-wide single sign-on platform.
 
-This repository contains a re-factored version of the [NITRO-Competitions](https://github.com/NUBIC/nitro-competitions) original code base which is currently running at Northwestern University Clinical and Translational Institute (NUCATS). It is work-progress, and intended primarily to enable a cloud-based implementtion in support of the [CTSA National Center for Data to Health (CD2H)](https://ctsa.ncats.nih.gov/cd2h/) operations. 
+This repository contains a re-factored version of the [NITRO-Competitions](https://github.com/NUBIC/nitro-competitions) original code base which is currently running at Northwestern University Clinical and Translational Institute (NUCATS).
+With the support of the [CTSA National Center for Data to Health (CD2H)](https://ctsa.ncats.nih.gov/cd2h/), the new application supports a cloud-based implementation and SAML authentication.
 
+# Prerequisites
+
+Installing Competitions requires certain configurations.
+  * A shibboleth IDP for these configuration
```
Comment (jseraf) [inline README.md:11]: This sentence seems to be missing something. Also, want to briefly include something related to #518 here (obv not about Okta).

File: config/competitions_config.example.yml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  # See https://github.com/apokalipto/devise_saml_authenticatable for more information.
+  saml:
+    saml_default_user_key: 'email' # The user will be looked up by this key
+    # The following maps the IDP SAML attribute's names(left)
+    #  with the User model attributes.
+    attribute_map:  { "email": "email",
+                      "firstName": "first_name",
+                      "lastName": "last_name",
+                      "upn": "upn"
+                    }
```
Comment (jseraf) [inline config/competitions_config.example.yml:26]: Do the `""` here have significance? Everything else is in `''`

File: config/application.rb [hunk 7]
```diff
@@ -26,3 +26,11 @@ class Application < Rails::Application
     config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '**', '*.{rb,yml}')]
   end
 end
+
+competitions_config = File.join(Rails.root, 'config', 'competitions_config.yml')
+
+if File.exists?(competitions_config)
+  COMPETITIONS_CONFIG = ActiveSupport::HashWithIndifferentAccess.new(YAML.load(File.open(competitions_config)))[Rails.env.to_sym]
+else
+  Rails.logger.info("Warning: Competitions config file is missing. (#{competitions_config})")
```
Comment (jseraf) [inline config/application.rb:35]: `Rails.logger.error` or `.fatal` rather than `.info`?

File: config/environments/development.rb [hunk 8]
```diff
@@ -65,13 +65,21 @@
...
+  config.action_mailer.raise_delivery_errors = COMPETITIONS_CONFIG[:mailer][:raise_delivery_errors]
+
   # Mailcatcher - download at https://mailcatcher.me/
-  config.action_mailer.delivery_method = :smtp
-  config.action_mailer.smtp_settings = { address: '127.0.0.1', port: 1025 }
-  config.action_mailer.raise_delivery_errors = true
-  config.subdomain = ''
-  config.app_domain = 'localhost'
-  config.action_mailer.default_url_options = { host: config.app_domain, port: 3000 }
+config.subdomain = COMPETITIONS_CONFIG[:subdomain]
```
Comment (jseraf) [inline config/environments/development.rb:74]: indent

File: config/competitions_config.example.yml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                    want_assertions_encrypted: false,
+                    want_name_id: false,
+                    metadata_signed: false,
+                    embed_sign: false,
+                    digest_method: "http://www.w3.org/2000/09/xmldsig#sha1",
+                    signature_method: "http://www.w3.org/2000/09/xmldsig#rsa-sha1",
+                    check_idp_cert_expiration: false,
+                    check_sp_cert_expiration: false }
+    # NOTE: Update idp_cert in config/secrets.yml.
+#Test should be included and configured in local environment.
```
Comment (jseraf) [inline config/competitions_config.example.yml:51]: Based on Teams conversation, clarify as needed?

File: config/storage.example.yml [hunk 10]
```diff
@@ -6,13 +6,14 @@ local:
...
-amazon:
-  service: S3
-  access_key_id: <%= Rails.application.credentials.dig(Rails.env.to_sym, :aws_access_key_id) %>
-  secret_access_key: <%= Rails.application.credentials.dig(Rails.env.to_sym, :aws_secret_access_key) %>
-  region: <%= Rails.application.credentials.dig(Rails.env.to_sym, :aws_region) %>
-  bucket: <%= Rails.application.credentials.dig(Rails.env.to_sym, :aws_bucket) %>
+# amazon:
+#   service: S3
+#   access_key_id: <%= Rails.application.credentials.dig(:aws, :access_key_id) %>
+#   secret_access_key: <%= Rails.application.credentials.dig(:aws, :secret_access_key) %>
```
Comment (jseraf) [inline config/storage.example.yml:14]: Should this be `secrets` rather than `Rails.application.credentials`

File: README.md [hunk 11]
```diff
@@ -2,5 +2,82 @@
...
+  * Storage should be configured locally or with AWS.
+    - It is possible to user other services, but more changes will be necessary.
+  * Mailcatcher(https://mailcatcher.me) is used in the development environment for emails.
+  * Postgres database
+  * RVM installed on server
+  * Capistrano for deployment
```
Comment (jseraf) [inline README.md:17]: I wonder if we just say Capistrano deployment (with RVM) is configurable out of the box? Someone may want to use something else.

File: lib/devise_saml_authenticatable/competitions_attribute_map_resolver.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+module DeviseSamlAuthenticatable
+  class CompetitionsAttributeMapResolver < DeviseSamlAuthenticatable::DefaultAttributeMapResolver
+    def attribute_map
+      COMPETITIONS_CONFIG[:saml][:attribute_map]
+    end
```
Comment (fanglinnw) [inline lib/devise_saml_authenticatable/competitions_attribute_map_resolver.rb:5]: I suggest that you add a unit test for this hack to make sure upgrading the gem doesn't break it.

Comment (fanglinnw) [reply lib/devise_saml_authenticatable/competitions_attribute_map_resolver.rb:5]: sorry I thought this was the hack John was talking about. I was confused. This is fine. I think integration test can cover this. Ignore my previous comment.

File: config/environments/production.example.rb [hunk 13]
```diff
@@ -72,15 +72,16 @@
...
 
-  config.action_mailer.default_url_options = Rails.application.credentials.dig(Rails.env.to_sym, :default_url_options)
-
-  config.subdomain = Rails.application.credentials.dig(Rails.env.to_sym, :subdomain)
-  config.app_domain = Rails.application.credentials.dig(Rails.env.to_sym, :app_domain)
-  config.action_mailer.delivery_method = Rails.application.credentials.dig(Rails.env.to_sym, :delivery_method).to_sym
-  config.action_mailer.perform_deliveries = Rails.application.credentials.dig(Rails.env.to_sym, :perform_deliveries)
-  config.action_mailer.smtp_settings = Rails.application.credentials.dig(Rails.env.to_sym, :smtp_settings)
+  config.subdomain = COMPETITIONS_CONFIG[:subdomain]
+  config.app_domain = COMPETITIONS_CONFIG[:app_domain]
```
[the remark this answers is not in the document]
Comment (mattbaumann1) [reply config/environments/production.example.rb:78]: Each environment has it's own config file stored on the server. so it pulls from the same place but the value is different based on environment.

**Pull request merged**